javatests/arcs/android/integration/policy/PolicyTest.kt

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2021 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.android.integration.policy
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import arcs.android.integration.IntegrationEnvironment
+import arcs.core.data.RawEntity
+import arcs.core.host.toRegistration
+import arcs.core.storage.StorageKey
+import arcs.core.storage.referencemode.ReferenceModeStorageKey
+import arcs.core.util.testutil.LogRule
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withTimeout
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@OptIn(ExperimentalCoroutinesApi::class, kotlin.time.ExperimentalTime::class)
+@RunWith(AndroidJUnit4::class)
+class PolicyTest {
+
+  @get:Rule
+  val log = LogRule()
+
+  /** Note, a new IntegrationEnvironment is allocated for each test method. */
+  @get:Rule
+  val env = IntegrationEnvironment()
+
+  /**
+   * Scenario: A policy-compliant recipe with @persistent handles egresses data
+   * and stores ingress-restricted values to disk that is never deleted.
+   */
+  @Test
+  fun persistentHandlesWithEgress_StoresIngressRestrictedValues_neverDeleted() = runBlocking {
+    env.addNewHostWith(
+      ::IngressThing.toRegistration(),
+      ::EgressAB.toRegistration()
+    )
+
+    val arc = env.startArc(PersistsEgressesPlan)
+
+    // When the Arc is run...
+    env.waitForIdle(arc)
+
+    val ingest = env.getParticle<IngressThing>(arc)
+    val egressAB = env.getParticle<EgressAB>(arc)
+
+    withTimeout(30000) {
+      ingest.storeFinished.join()
+      egressAB.handleRegistered.join()
+    }
+
+    // Then data with fields Thing {a, b} will be egressed
+    assertThat(egressAB.outputForTest).hasSize(6)
+
+    val startingKey = (egressAB.handles.output.getProxy().storageKey as ReferenceModeStorageKey)
+      .storageKey
+
+    // And only Thing {a, b} is written to storage
+    assertOnlySingletonAbIsPersisted(startingKey)
+
+    env.stopArc(arc)
+
+    // And Thing {a, b} data persists after the arc is run
+    assertOnlySingletonAbIsPersisted(startingKey)
+
+    env.stopRuntime()
+
+    // And Thing {a, b} data persists after runtime ends
+    assertOnlySingletonAbIsPersisted(startingKey)
+  }
+
+  private suspend fun assertOnlySingletonAbIsPersisted(startingKey: StorageKey) {
```

Comment:
I would have made this mirror some of the arguments of `assertRawEntityHasFields` method as follows:
```
private suspend fun assertDatabaseEntityHasFields(
   startingKey: StorageKey,
   schema: ...,
   singletons: Collection<String> = emptySet(),
   collections: Collection<String> = emptySet()
)
```
 so that you can do invoke it as follows:
```
assertDatabaseHasFields(startingKey, AbstractIngressThing.Thing.SCHEMA, setOf("a", "b"))
```

Reply:
I'm currently trying not to generalize. I'm trying not to create abstractions to solve problems that I'll have in the future, with the intention to rewrite once I discover a useful pattern. This approach is inspired by [this article](https://programmingisterrible.com/post/176657481103/repeat-yourself-do-more-than-one-thing-and). What do you think?

Reply:
It is fine for now, but I think you will need something like this in the next couple of tests anyway.